src/AtomicLaunch.sol

```diff
@@ -0,0 +1,242 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.12;
+
+import "openzeppelin-contracts/utils/structs/EnumerableSet.sol";
+
+import {ChainlinkUtils} from "./oracles/ChainlinkUtils.sol";
+import {CtdlWbtcCurveV2Provider} from "./oracles/CtdlWbtcCurveV2Provider.sol";
+import {CtdlAssetChainlinkProvider} from "./oracles/CtdlAssetChainlinkProvider.sol";
+import {CtdlBtcChainlinkProvider} from "./oracles/CtdlBtcChainlinkProvider.sol";
+import {CtdlEthChainlinkProvider} from "./oracles/CtdlEthChainlinkProvider.sol";
+import {CtdlWibbtcLpVaultProvider} from "./oracles/CtdlWibbtcLpVaultProvider.sol";
+
+import "./interfaces/erc20/IERC20.sol";
+import "./interfaces/curve/ICurvePoolFactory.sol";
+import "./interfaces/curve/ICurvePool.sol";
+import "./interfaces/chainlink/IAggregatorV3Interface.sol";
+
+contract AtomicLaunch is ChainlinkUtils {
+    using EnumerableSet for EnumerableSet.AddressSet;
+
+    address public citadel;
+    address public wbtc;
+
+    address public constant WBTC_BTC_PRICE_FEED =
+        0xfdFD9C85aD200c506Cf9e21F1FD8dd01932FBB23;
+    address public constant BTC_ETH_PRICE_FEED =
+        0xdeb288F737066589598e9214E782fa5A8eD689e8;
+
+    address public constant ETH_USD_PRICE_FEED =
+        0x5f4eC3Df9cbd43714FE2740f5E3616155c5b8419;
+
+    address public constant WIBBTC_LP_VAULT =
+        0xaE96fF08771a109dc6650a1BdCa62F2d558E40af;
+
+    address public constant FRAX_ETH_PRICE_FEED =
+        0x14d04Fff8D21bd62987a5cE9ce543d2F1edF5D3E;
+    address public constant FRAX_USD_PRICE_FEED =
+        0xB9E1E3A9feFf48998E45Fa90847ed4D467E8BcfD;
+
+    address public constant USDC_ETH_PRICE_FEED =
+        0x986b5E1e1755e3C2440e960477f25201B0a8bbD4;
+    address public constant USDC_USD_PRICE_FEED =
+        0x8fFfFfd4AfB6115b954Bd326cbe7B4BA576818f6;
+
+    address public constant CVX_ETH_PRICE_FEED =
+        0xC9CbF687f43176B302F03f5e58470b77D07c61c6;
+    address public constant CVX_USD_PRICE_FEED =
+        0xd962fC30A72A84cE50161031391756Bf2876Af5D;
+
+    address public constant BADGER_ETH_PRICE_FEED =
+        0x58921Ac140522867bf50b9E009599Da0CA4A2379;
+    address public constant BADGER_USD_PRICE_FEED =
+        0x66a47b7206130e6FF64854EF0E1EDfa237E65339;
+
+    address public constant CURVE_POOL_FACTORY =
+        0xF18056Bbd320E96A48e3Fbf8bC061322531aac99;
+    uint256 public constant CITADEL_PRICE = 21;
+
+    IAggregatorV3Interface public wbtcUsdPriceFeed =
+        IAggregatorV3Interface(0xF4030086522a5bEEa4988F8cA5B36dbC97BeE88c);
+
+    /* ========== STATE VARIABLES ========== */
+    address public governance;
+
+    EnumerableSet.AddressSet internal _oracles;
```

Comment:
We can probably replace this with a simple array

Reply:
thanks for the suggestion.

Notes from @shuklaayush for the reasoning of gas saving:

```
- Saves a storage slot for each element
- One sstore is 20000 gas. 12 * 20000 = 240000 gas.
```